crates/optimism/flashblocks/src/service.rs

```diff
@@ -177,14 +211,43 @@ impl<
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let this = self.get_mut();
 
-        match this.rx.poll_next_unpin(cx) {
-            Poll::Ready(Some(Ok(flashblock))) => {
-                this.add_flash_block(flashblock);
-                Poll::Ready(Some(this.execute()))
+        // Consume new flashblocks while they're ready
+        while let Poll::Ready(Some(result)) = this.rx.poll_next_unpin(cx) {
+            match result {
+                Ok(flashblock) => this.add_flash_block(flashblock),
+                Err(err) => return Poll::Ready(Some(Err(err))),
+            }
+        }
+
+        // Execute block if there are flashblocks but no last pending block
+        let new_block = if this.current.is_none() && !this.blocks.is_empty() {
+            match this.execute() {
+                Ok(block) => Some(block),
+                Err(err) => return Poll::Ready(Some(Err(err))),
+            }
+        } else {
+            None
+        };
+
+        let changed = new_block.is_some();
+
+        if let Some(new_block) = new_block {
+            this.current = Some(new_block);
+        }
+
+        // Verify that pending block is following up to the canonical state
+        match this.verify_pending_block_integrity(cx) {
+            Ok(Some(false)) => Poll::Ready(None), // Integrity check failed, erase last block
+            Ok(Some(true)) => {
+                // Integrity check is OK
+                if changed {
+                    Poll::Ready(this.current.clone().map(Ok)) // Output new block
+                } else {
+                    Poll::Pending // Keep last block
+                }
             }
-            Poll::Ready(Some(Err(err))) => Poll::Ready(Some(Err(err))),
-            Poll::Ready(None) => Poll::Ready(None),
-            Poll::Pending => Poll::Pending,
+            Ok(None) => Poll::Pending, // Cannot check integrity, tip or block is missing
+            Err(err) => Poll::Ready(Some(Err(err))),
```

Comment:
I'm realizing that there's now an architectural issue because this stream only emits new blocks, so we can't actually clear the pending block, so three solutions, either we move the watch channel in here and make this type the future, or we move the canonsubscription outside, or we make enum item for this stream, which doesnt seem ideal.